server/scripts/migrateIdentities.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * This script "migrates" identities from the chain into the database, by first
+ * querying all addresses, and then attempting to fetch corresponding identities
+ * from the chain, writing the results back into the database.
+ */
+
+import _ from 'underscore';
+import { SubstrateTypes, SubstrateEvents } from '@commonwealth/chain-events';
+import { OffchainProfileAttributes, OffchainProfileInstance } from '../models/offchain_profile';
+import IdentityEventHandler from '../eventHandlers/identity';
+
+import { factory, formatFilename } from '../../shared/logging';
+const log = factory.getLogger(formatFilename(__filename));
+
+export async function migrateIdentities(models, chain?: string): Promise<void> {
+  // 1. fetch the node and url of supported/selected chains
+  log.info('Fetching node info for identity migrations...');
+  const nodes = [];
+  if (!chain) {
+    // query one node for each supported chain
+    for (const supportedChain of SubstrateTypes.EventChains) {
+      // eslint-disable-next-line no-await-in-loop
+      nodes.push(await models.ChainNode.findOne({
+        where: { chain: supportedChain }
+      }));
+    }
+  } else {
+    // query one node for provided chain
+    nodes.push(await models.ChainNode.findOne({
+      where: { chain }
+    }));
+  }
+  if (!nodes) {
+    throw new Error('no nodes found for identity migration');
+  }
+
+  // 2. for each node, fetch and migrate identities
+  for (const node of nodes) {
+    // 2a. query all available addresses in the db on the chain
+    log.info(`Querying all profiles on chain ${node.chain}...`);
+    const profiles: OffchainProfileInstance[] = await models.OffchainProfile.findAll({
+      include: [{
+        model: models.Address,
+        where: {
+          chain: node.chain,
+        },
+        required: true,
+      }]
+    });
+    const addresses = profiles.map((p) => p.Address.address);
+
+    // 2b. connect to chain and query all identities of found addresses
+    log.info(`Fetching identities on chain ${node.chain} at url ${node.url}...`);
+    let nodeUrl = node.url;
+    const hasProtocol = nodeUrl.indexOf('wss://') !== -1 || nodeUrl.indexOf('ws://') !== -1;
+    nodeUrl = hasProtocol ? nodeUrl.split('://')[1] : nodeUrl;
+    const isInsecureProtocol = nodeUrl.indexOf('kusama-rpc.polkadot.io') === -1
+      && nodeUrl.indexOf('rpc.polkadot.io') === -1;
+    const protocol = isInsecureProtocol ? 'ws://' : 'wss://';
+    if (nodeUrl.indexOf(':9944') !== -1) {
+      nodeUrl = isInsecureProtocol ? nodeUrl : nodeUrl.split(':9944')[0];
+    }
+    nodeUrl = protocol + nodeUrl;
+    const provider = await SubstrateEvents.createProvider(nodeUrl);
+    const api = await SubstrateEvents.createApi(provider, node.chain).isReady;
+    const fetcher = new SubstrateEvents.StorageFetcher(api);
+    const identityEvents = await fetcher.fetchIdentities(addresses);
+
+    // 2c. remove all existing identities from profiles
+    await Promise.all(profiles.map(async (p) => {
```

Comment:
Is there a reason why we would remove them all first and not just update the ones in the latest fetch? Is this what you were mentioning with regards to deleting non-existent identities.. and is there no way we catch an event where you delete your identity?

Are we even sure you CAN delete an identity? or only just change it do something with zero'd values?

Reply:
We do catch events where you delete your identity. However, this is for when we've been offline for a while. We don't _have_ to delete them all, but it's simpler logic than comparing the state of existing identities with the set of newly-fetched identities, and then diffing them and deleting only the ones that got removed since last fetch.

Reply:
To be clear: the migration is intended to be run once, so this feature is actually not _technically_ needed, but in the event that we _do_ run it multiple times, I figured I'd put it in.